src/seeds.harvest.cpp

```diff
@@ -905,3 +864,95 @@ ACTION harvest::setorgtxpt(name organization, uint64_t tx_points) {
   } 
 
 }
+
+void harvest::calcmqevs () {
+  require_auth(get_self());
+  
+  uint64_t day = utils::get_beginning_of_day_in_seconds();
+  uint64_t cutoff = day - utils::moon_cycle;
+  
+  qev_tables qevs(contracts::history, contracts::history.value);
+  check(qevs.begin() != qevs.end(), "The qevs table for " + contracts::history.to_string() + " is empty");
+
+  auto qitr = qevs.rbegin();
+  uint64_t total_volume = 0;
+
+  while (qitr != qevs.rend() && qitr -> timestamp >= cutoff) {
+    total_volume += qitr -> qualifying_volume;
+    qitr++;
+  }
+
+  circulating_supply_table c = circulating.get();
+
+  auto mqitr = monthlyqevs.find(day);
+  
+  if (mqitr != monthlyqevs.end()) {
+    monthlyqevs.modify(mqitr, _self, [&](auto & item){
+      item.qualifying_volume = total_volume;
+      item.circulating_supply = c.circulating;
+    });
+  } else {
+    monthlyqevs.emplace(_self, [&](auto & item){
+      item.timestamp = day;
+      item.qualifying_volume = total_volume;
+      item.circulating_supply = c.circulating;
+    });
+  }
+}
+
+void harvest::testcalcmqev (uint64_t day, uint64_t total_volume, uint64_t circulating) {
+  require_auth(get_self());
+  
+  auto mqitr = monthlyqevs.find(day);
+  
+  if (mqitr != monthlyqevs.end()) {
+    monthlyqevs.modify(mqitr, _self, [&](auto & item){
+      item.qualifying_volume = total_volume;
+      item.circulating_supply = circulating;
+    });
+  } else {
+    monthlyqevs.emplace(_self, [&](auto & item){
+      item.timestamp = day;
+      item.qualifying_volume = total_volume;
+      item.circulating_supply = circulating;
+    });
+  }
+}
+
+
+void harvest::calcmintrate () {
+  require_auth(get_self());
+
+  uint64_t day = utils::get_beginning_of_day_in_seconds();
+  auto previous_day_temp = eosio::time_point_sec((day - (3 * utils::moon_cycle)) / 86400 * 86400);
+  uint64_t previous_day = previous_day_temp.utc_seconds;
+
+  auto current_qev_itr = monthlyqevs.get(day, "no monthly qev found");
+  auto previous_qev_itr = monthlyqevs.get(previous_day, "no previous monthly qev found");
```

Comment:
When we are starting up we may not have a 3 quarter lookback. I guess it's ok for the action to do nothing in that case, but wouldn't we stop the scheduler with an error?

Reply:
Yes, I will replace it with a .find